### PR TITLE
Deploy and Set up quiz in one task

### PR DESCRIPTION
**Pull request by ZigaMr**
Updated hardhat.config to include quiz deploy and entire setup in a single task.

### CODE REVIEW

**Review by matevz — COMMENTED** [review]
Can you remove 3a389f015187f876ba3336635530a7bc26dddaac from this PR so it can be cleanly merged to the main branch?

**Comment by ZigaMr** [comment]
Removed the duplicated commit.

File: backend/hardhat.config.ts [hunk 1]
```diff
@@ -260,6 +260,91 @@ task('setGaslessKeyPair')
...
+  .addOptionalParam('questionsfile', 'File containing questions in JSON format', 'test-questions.json')
+  .addOptionalParam('couponsfile', 'File containing coupons, one per line', 'test-coupons.txt')
+  .addOptionalParam('reward', 'Reward in ROSE', '2.0')
+  .addOptionalParam('gaslessaddress', 'Payer address for gasless transactions')
+  .addOptionalParam('gaslesssecret', 'Payer secret key for gasless transactions')
+  .addOptionalParam('fundamount', 'Amount in ROSE to fund the contract', '100')
+  .addOptionalParam('fundgaslessamount', 'Amount in ROSE to fund the gasless account', '10')
+  .addOptionalParam('contractaddress', 'Contract address for status check')
+  .setAction(async (args, hre) => {
+    // Run compilation
```
Comment (matevz) [inline backend/hardhat.config.ts]: This is obvious.
```suggestion
```

File: backend/hardhat.config.ts [hunk 2]
```diff
@@ -260,6 +260,91 @@ task('setGaslessKeyPair')
...
+    await hre.run('compile');
+
+    // For deployment unwrap the provider to enable contract verification.
+    const uwProvider = new JsonRpcProvider(hre.network.config.url);
+    // const accounts = await hre.ethers.getSigners();
```
Comment (matevz) [inline backend/hardhat.config.ts]: Leftover?

File: backend/hardhat.config.ts [hunk 3]
```diff
@@ -260,6 +260,91 @@ task('setGaslessKeyPair')
...
+
+    // Deploy Quiz contract
```
Comment (matevz) [inline backend/hardhat.config.ts]: Add full stops at the end of one-liner comments ;)

```suggestion
    // Deploy Quiz contract.
```

File: backend/hardhat.config.ts [hunk 4]
```diff
@@ -260,6 +260,91 @@ task('setGaslessKeyPair')
...
+    const Quiz = await hre.ethers.getContractFactory('Quiz', new hre.ethers.Wallet(accounts[0], uwProvider));
```
Comment (matevz) [inline backend/hardhat.config.ts]: Let's not duplicate any of this code. Move these blocks into separate functions (deploy, addQuestions, addCoupons etc.) and call those either from the `deployAndSetup` or separately.

File: README.md [hunk 5]
```diff
@@ -113,6 +113,13 @@ Checklist after deploying a production-ready quiz:
    npx hardhat getCoupons 0x385cAE1F3afFC50097Ca33f639184f00856928Ff --network sapphire-testnet
    ```
 
+### Deploy and setup quiz in with a single task
+
+You can also run the entire quiz with ```deployAndSetupQuiz```. This script handles all the above steps, you only need to provide the PRIVATE_KEY environment variable.
```
Comment (matevz) [inline README.md]: Please wrap to 80 characters.

File: README.md [hunk 6]
```diff
@@ -113,6 +113,13 @@ Checklist after deploying a production-ready quiz:
...
+   ```shell
+   npx hardhat deployAndSetupQuiz --gaslessaddress <gaslessAddress> --gaslesssecret <gaslessSecret> --network sapphire-testnet
+   ```
```
Comment (matevz) [inline README.md]: The markdown format prefers no indention and a new line before the code block. Also, let's document the simple-most deployAndSetupQuiz flavor for a functioning quiz and then below optional parameters (gasless etc.).

````suggestion

```shell
npx hardhat deployAndSetupQuiz --questionsFile questions.json --couponsFile coupons.json --network sapphire-testnet
```   
````

File: README.md (hunk 5 above)
Comment (ZigaMr) [reply README.md]: Shortened the description, now wraps to below 80 character limit.

File: README.md (hunk 6 above)
Comment (ZigaMr) [reply README.md]: Added basic deployAndSetupQuiz example and documented optional parameters below.

File: backend/hardhat.config.ts (hunk 4 above)
Comment (ZigaMr) [reply backend/hardhat.config.ts]: Agreed, added functions for duplicated code.

File: backend/hardhat.config.ts (hunk 3 above)
Comment (ZigaMr) [reply backend/hardhat.config.ts]: Updated the comments and removed unnecessary leftover code.

File: README.md [hunk 7]
```diff
@@ -113,6 +113,26 @@ Checklist after deploying a production-ready quiz:
...
+
+1. **`--questionsfile`**: A file containing questions in JSON format. Default value is `test-questions.json`.
+2. **`--couponsfile`**: A file containing coupons, one per line. Default value is `test-coupons.txt`.
+3. **`--reward`**: The reward in ROSE. Default value is `2.0`.
+4. **`--gaslessaddress`**: The payer address for gasless transactions.
+5. **`--gaslesssecret`**: The payer secret key for gasless transactions.
+6. **`--fundamount`**: The amount in ROSE to fund the contract. Default value is `100`.
+7. **`--fundgaslessamount`**: The amount in ROSE to fund the gasless account. Default value is `10`.
+8. **`--contractaddress`**: The contract address for status check.
+
```
Comment (matevz) [inline README.md:134]:
```suggestion
```

File: backend/hardhat.config.ts [hunk 8]
```diff
@@ -10,7 +10,71 @@ import 'hardhat-watcher';
 import { TASK_COMPILE } from 'hardhat/builtin-tasks/task-names';
 import { HardhatUserConfig, task } from 'hardhat/config';
 import 'solidity-coverage';
-import {ethers} from "hardhat";
+
+
```
Comment (matevz) [inline backend/hardhat.config.ts:14]:
```suggestion
```

File: backend/hardhat.config.ts [hunk 9]
```diff
@@ -244,22 +274,50 @@ task('reclaimFunds')
...
 
-    let quiz = await hre.ethers.getContractAt('Quiz', args.address);
 
-    const nonce = await hre.ethers.provider.getTransactionCount(args.payerAddress);
-    const tx = await quiz.setGaslessKeyPair(args.payerAddress, args.payerSecret, nonce);
-    const receipt = await tx.wait();
-    console.log(`Successfully set gasless keypair to ${args.payerAddress}, secret ${args.payerSecret} and nonce ${nonce}. Transaction hash: ${receipt!.hash}`);
+// Deploy and setup Quiz contract.
+task('deployAndSetupQuiz')
+  .addOptionalParam('questionsfile', 'File containing questions in JSON format', 'test-questions.json')
```
Comment (matevz) [inline backend/hardhat.config.ts]: Use camel case here and in the rest of optional params. Hardhat will then automatically replace it to `--questions-file`. Don't forget to update the readme accordingly.

```suggestion
  .addOptionalParam('questionsFile', 'File containing questions in JSON format', 'test-questions.json')
```

File: backend/hardhat.config.ts [hunk 10]
```diff
@@ -244,22 +274,50 @@ task('reclaimFunds')
...
+    await setGaslessKeyPair(quiz, args.gaslessaddress, args.gaslesssecret, nonce);
+    await fundGaslessAccount(hre, args.gaslessaddress, args.fundgaslessamount);
+    if (args.contractaddress) {
+      await hre.run('status', { address: args.contractaddress });
+    } else {
+      await hre.run('status', { address: await quiz.getAddress() });
+    }
   });
 
+
```
Comment (matevz) [inline backend/hardhat.config.ts]:
```suggestion
```

**Review by matevz — COMMENTED** [review]
Some minor nits, otherwise looks good.

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1,2 @@
+node_modules
```
Comment (matevz) [inline .gitignore:1]: There are .gitignore in the backend/ and frontend/ folders already filtering out node_modules, so there is no need for another one here.

Comment (ZigaMr) [reply .gitignore:1]: Removed duplicated .gitignore files and updated the parent file.

**Review by matevz — APPROVED** [review]

**Pull request merged**